api/apps/api/src/modules/geo-features/geo-features.service.ts

```diff
@@ -379,10 +420,51 @@ export class GeoFeaturesService extends AppBaseService<
     return extendedResult;
   }
 
+  public async uploadFeatureFromShapefile(
+    projectId: string,
+    dto: UploadShapefileDTO,
+    shapefile: Express.Multer.File,
+  ): Promise<Either<typeof onlyFeatureCollectionShapefileSupported, void>> {
+    await ensureShapefileHasRequiredFiles(shapefile);
+
+    let data: GeoJSON;
+    try {
+      const geoJSON = await this.shapefileService.transformToGeoJson(
+        shapefile,
+        {
+          allowOverlaps: true,
+        },
+      );
+      data = geoJSON.data;
+    } catch (e) {
+      throw new BadRequestException(e);
+    }
+
+    if (!isFeatureCollection(data)) {
+      return left(onlyFeatureCollectionShapefileSupported);
+    }
+
+    this.createFeaturesForShapefileAsync(projectId, dto, data.features);
+
+    return right(void 0);
+  }
+
+  public createFeaturesForShapefileAsync(
+    projectId: string,
+    data: UploadShapefileDTO,
+    featureDatas: Record<string, any>[],
+  ) {
+    try {
+      setTimeout(async () => {
+        await this.createFeaturesForShapefile(projectId, data, featureDatas);
+      });
+    } catch (e) {}
```

Comment:
Ah yes, this is because in order to make this run asynchronously, by wrapping the execution in a setTimeout function, an error thrown inside it completely slips by the global exception filter, as the function gets executed from the global context.
I found this while tweaking the CSV tests that checked for failures, whenever an error was thrown it would terminate the test prematurely.

https://nodejs.org/dist/latest-v7.x/docs/api/errors.html#errors_node_js_style_callbacks
https://stackoverflow.com/questions/41431605/how-to-handle-errors-from-settimeout-in-javascript